src/helpers/figures/charts/waterfall_chart.ts

```diff
@@ -0,0 +1,357 @@
+      ticks: {
+        color: fontColor,
+        callback: (value) => {
+          value = Number(value);
+          if (isNaN(value)) return value;
+          return formatValue(value, {
+            locale,
+            format: !format && Math.abs(value) > 1000 ? "#,##" : format,
+          });
+        },
+      },
+      grid: {
+        lineWidth: (context) => {
+          return context.tick.value === 0 ? 2 : 1;
+        },
+      },
+    },
+  };
+  config.options.plugins!.tooltip = {
+    callbacks: {
+      label: function (tooltipItem) {
+        const [lastValue, currentValue] = tooltipItem.raw as [number, number];
+        const yLabel = currentValue - lastValue;
+        const dataSeriesIndex = Math.floor(tooltipItem.dataIndex / labels.length);
+        const dataSeriesLabel = dataSeriesLabels[dataSeriesIndex];
+        const toolTipFormat = !format && Math.abs(yLabel) > 1000 ? "#,##" : format;
+        const yLabelStr = formatValue(yLabel, { format: toolTipFormat, locale });
+        return dataSeriesLabel ? `${dataSeriesLabel}: ${yLabelStr}` : yLabelStr;
+      },
+    },
+  };
+  config.options.plugins!.waterfallLinesPlugin = { showConnectorLines: chart.showConnectorLines };
+
+  return config;
+}
+
+export function createWaterfallChartRuntime(
+  chart: WaterfallChart,
+  getters: Getters
+): WaterfallChartRuntime {
+  const labelValues = getChartLabelValues(getters, chart.dataSets, chart.labelRange);
+  let labels = labelValues.formattedValues;
+  let dataSetsValues = getChartDatasetValues(getters, chart.dataSets);
+  if (
+    chart.dataSetsHaveTitle &&
+    dataSetsValues[0] &&
+    labels.length > dataSetsValues[0].data.length
+  ) {
+    labels.shift();
+  }
+
+  ({ labels, dataSetsValues } = filterEmptyDataPoints(labels, dataSetsValues));
+  if (chart.aggregated) {
+    ({ labels, dataSetsValues } = aggregateDataForLabels(labels, dataSetsValues));
+  }
+  if (chart.showSubTotals) {
+    labels.push(_t("Subtotal"));
+  }
+
+  const dataSetFormat = getChartDatasetFormat(getters, chart.dataSets);
+  const locale = getters.getLocale();
+  const dataSeriesLabels = dataSetsValues.map((dataSet) => dataSet.label);
+  const config = getWaterfallConfiguration(chart, labels, dataSeriesLabels, {
+    format: dataSetFormat,
+    locale,
+  });
+  config.type = "bar";
+
+  const negativeColor = chart.negativeValuesColor || CHART_WATERFALL_NEGATIVE_COLOR;
+  const positiveColor = chart.positiveValuesColor || CHART_WATERFALL_POSITIVE_COLOR;
+  const subTotalColor = chart.subTotalValuesColor || CHART_WATERFALL_SUBTOTAL_COLOR;
+
+  const backgroundColor: Color[] = [];
+  const datasetValues: Array<[number, number]> = [];
+  const dataset: ChartDataset = {
+    label: "",
+    data: datasetValues,
+    backgroundColor,
+  };
+  const labelsWithSubTotals: string[] = [];
+  let lastValue = 0;
+  for (const dataSetsValue of dataSetsValues) {
+    for (let i = 0; i < dataSetsValue.data.length; i++) {
```

Comment:
`for (const [i, data] of dataSetsValue.entries()) {` ? Or should we avoid this kind of stuff ?

Reply:
We should avoid using `entires()`, because it allocates new arrays. Here if we create a chart on a whole column, that's like 10K arrays allocated for nothing